atmel-samd/tools/update_asf.py

```diff
@@ -16,39 +18,37 @@
             sys.exit(1)
         with open(filename, "wb") as out:
             out.write(r.content)
-    z = zipfile.ZipFile(filename)
-    z.extractall("asf4/" + chip)
-    print(z)
 
-    # delete the zip on success
-    #os.remove(filename)
+    # Extract to a temporary location and patch it before replacing the existing location.
+    z = zipfile.ZipFile(filename)
+    tmp_dir = "asf4/" + chip + "_vanilla"
+    z.extractall(tmp_dir)
 
-# Remove LITTLE_ENDIAN define from ASF because it conflicts with GCC.
-for fn in os.listdir("asf4/samd51/include"):
-    contents = []
-    path = "asf4/samd51/include/" + fn
-    if not os.path.isfile(path):
-        continue
-    with open(path, "r") as f:
-        for line in f:
-            if line.startswith("#define LITTLE_ENDIAN"):
-                contents.append("//" + line)
-            else:
-                contents.append(line)
-    with open(path, "w") as f:
-        for line in contents:
-            f.write(line)
+    # Remove all carriage returns.
+    for dirpath, dirnames, filenames in os.walk(tmp_dir):
+        for fn in filenames:
+            fn = os.path.join(dirpath, fn)
+            subprocess.run(["sed", "-i", "s/\r//g", fn])
 
-# Replace ASF's assert with asf_assert so it doesn't conflict with the C macro.
+    # Move files to match SAMD51 structure.
+    if chip == "samd21":
+        shutil.move("asf4/samd21_vanilla/samd21a/include", "asf4/samd21_vanilla")
+        shutil.move("asf4/samd21_vanilla/samd21a/gcc/gcc", "asf4/samd21_vanilla/gcc")
+        shutil.move("asf4/samd21_vanilla/samd21a/gcc/system_samd21.c", "asf4/samd21_vanilla/gcc")
 
-for chip in ["samd21", "samd51"]:
-    contents = []
-    path = "asf4/" + chip + "/hal/utils/include/utils_assert.h"
-    with open(path, "r") as f:
-        for line in f:
-            contents.append(line.replace(" assert(", " asf_assert("))
-    with open(path, "w") as f:
-        for line in contents:
-            f.write(line)
+    ok = True
+    for patch in os.listdir("asf4/patches/" + chip):
+        patch = "patches/" + chip + "/" + patch
+        print(patch)
+        result = subprocess.run(["patch", "-l", "-F", "10", "-u", "-p", "1", "-d", tmp_dir, "-i", "../" + patch])
```

Comment:
Doesn't need to be changed, but consider using `subprocess.run("some {} command {}".format(...), shell=True)` to make the command line easier to change and read.